collectors/recon.go

```diff
@@ -0,0 +1,637 @@
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *ReconCollector) Collect(ch chan<- prometheus.Metric) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(c.tasks))
+	for name, task := range c.tasks {
+		go func(name string, task func(string, string, chan<- prometheus.Metric)) {
+			task(name, c.pathToExecutable, ch)
+			wg.Done()
+		}(name, task)
+	}
+	wg.Wait()
+}
+
+var (
+	clusterStorageUsedPercentByDiskDesc = prometheus.NewDesc(
+		"swift_cluster_storage_used_percent_by_disk",
+		"Percentage of storage used by a disk as reported by the swift-recon tool.",
+		[]string{"storage_ip", "disk"}, nil,
+	)
+	clusterStorageUsedPercentDesc = prometheus.NewDesc(
+		"swift_cluster_storage_used_percent",
+		"Percentage of storage used as reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+	clusterStorageUsedBytesDesc = prometheus.NewDesc(
+		"swift_cluster_storage_used_bytes",
+		"Used storage bytes as reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+	clusterStorageFreeBytesDesc = prometheus.NewDesc(
+		"swift_cluster_storage_free_bytes",
+		"Free storage bytes as reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+	clusterStorageCapacityBytesDesc = prometheus.NewDesc(
+		"swift_cluster_storage_capacity_bytes",
+		"Capacity storage bytes as reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+)
+
+var specialCharRx = regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+func reconDiskUsageTask(taskName, pathToReconExecutable string, ch chan<- prometheus.Metric) {
+	errCount := 0
+	cmdArgs := []string{"--diskusage", "--verbose"}
+	sendReconErrorCount(ch, cmdArgs, errCount)
+
+	result, err := getSwiftReconOutputPerHost(pathToReconExecutable, cmdArgs...)
+	if err != nil {
+		logg.Error("recon collector: %s: %v", taskName, err)
+		errCount++
+		sendReconErrorCount(ch, cmdArgs, errCount)
+		return
+	}
+
+	gotErr := false
+	for hostname, dataBytes := range result {
+		var disksData []struct {
+			Device  string `json:"device"`
+			Avail   int64  `json:"avail"`
+			Mounted bool   `json:"mounted"`
+			Used    int64  `json:"used"`
+			Size    int64  `json:"size"`
+		}
+		err = json.Unmarshal(dataBytes, &disksData)
+		if err != nil {
+			logg.Error("recon collector: %s: %s: %v", taskName, hostname, err)
+			gotErr = true
+			continue
+		}
+
+		var totalFree, totalUsed, totalSize int64
+		for _, disk := range disksData {
+			if !(disk.Mounted) {
+				continue
+			}
+
+			totalFree += disk.Avail
+			totalUsed += disk.Used
+			totalSize += disk.Size
+
+			// submit metrics by disk (only used percent here, which is the
+			// most useful for alerting)
+			device := specialCharRx.ReplaceAllLiteralString(disk.Device, "")
+			ch <- prometheus.MustNewConstMetric(
+				clusterStorageUsedPercentByDiskDesc,
+				prometheus.GaugeValue, float64(disk.Used)/float64(disk.Size),
+				hostname, device,
+			)
+		}
+
+		usedPercent := float64(totalUsed) / float64(totalSize)
```

Comment:
That's a fractional value, not a percentage value. Percentages are 0..100, but this ranges between 0 and 1.

Reply:
9c69349 does not fix this issue. The metric still has "percent" in its name and description.

Reply:
This is resolved after discussion in Slack: The legacy implementation emits this metric in the same way with the same name, so we'll keep it like it is.